### ISSUE TITLE
Bug with SERVICE directive

### ISSUE THREAD

**Issue by c-martinez** [ClioPatria/ClioPatria#54]
This query returns one result:

```
SELECT * WHERE
{
  SERVICE <http://dbpedia-live.openlinksw.com/sparql>
  {
    ?entity rdf:type <http://dbpedia.org/ontology/Person> .
    ?entity rdfs:label "Frida Kahlo"@en
  }
}
```

However, this doesn't seem to work when using `BIND` to pass the search term to the `SERVICE` directive:

```
SELECT * WHERE
{
  BIND( "Frida Kahlo"@en as ?term )
  SERVICE <http://dbpedia-live.openlinksw.com/sparql>
  {
    ?entity rdf:type <http://dbpedia.org/ontology/Person> .
    ?entity rdfs:label ?term
  }
}
```

I seem to be getting a `Type error: `rdf_object' expected, found `literal(lang(en,'Frida Kahlo'))' (a compound)` error.

Is this a bug? Or is there something wrong with my query? I have compared the behaviour with other SPARQL endpoints (Blazegraph & snorql) but second query seems to work ok on those.

**Comment by jrvosse** [comment]
On 06/27/2017 12:23 PM, Carlos Martinez wrote:
> This query returns one result:
>
> |SELECT * WHERE { SERVICE <http://dbpedia-live.openlinksw.com/sparql> { ?entity rdf:type <http://dbpedia.org/ontology/Person> . ?entity rdfs:label "Frida Kahlo"@en } } |
>
> However, this doesn't seem to work when using |BIND| to pass the search term to the |SERVICE| directive:
>
> |SELECT * WHERE { BIND( "Frida Kahlo"@en as ?term ) SERVICE <http://dbpedia-live.openlinksw.com/sparql> { ?entity rdf:type <http://dbpedia.org/ontology/Person> . ?entity rdfs:label ?term } } |
>
> I seem to be getting a |Type error:|rdf_object' expected, found |literal(lang(en,'Frida Kahlo'))' (a compound)| error.
>
> Is this a bug?

I think it is a bug. I was experiencing similar problems.

Jan, Wouter: I do not understand how the explicit mixing of rdf_db and rdf11 literals in ClioPatria/rdfql/sparql_runtime.pl is intended to work.
This module first imports rdf_db and then

:- use_module(library(semweb/rdf11), [rdf_lexical_form/2]).

This looks like the use of the new rdf11:rdf_lexical_form in "old" rdf_db code in the example above is intentional, e.g. in

binding_line(Name=Literal, Line) :-
	rdf_lexical_form(Literal, Lex), % THIS LINE expects Literal to be in the new "Frida Kahlo"@en format and throws the exception on the old literal(lang(en,'Frida Kahlo'))
	(   Lex = ^^(String,Type)
	->  format(string(Line), 'BIND(~w^^<~w> as ?~w)', [String, Type, Name])
	;   Lex = @(String,Lang)
	->  format(string(Line), 'BIND(~w@~w as ?~w)', [String, Lang, Name])
	).

Jan, is this the same "should we have a general solution to handle literals when mixing old rdf_db and new rdf11" discussion we had last week?


Thanks,

Jacco

**Comment by wouterbeek** [comment]
My personal preference is to move ClioPatria to rdf11 entirely and go for a point release that breaks compatibility.  (I believe this is TRT given the number of ClioPatria developers.)  It would be nice if we could also update the `rdf_load/[1,2]` parsers to generate rdf11 terms.

**Comment by JanWielemaker** [comment]
Fixed with b80a1f29c6bffd3c55ac7d580cd7ad916aa869a4, which also requires todays git version of SWI-Prolog as part of the patch is in SWI-Prolog's semweb package.
